### PR TITLE
netflow: Add observation domain and point to message

### PR DESCRIPTION
**Pull request by amorenoz**
The ObservationDomainID and ObservationPointID are two IPFIX fields that
identify the entity that is capturing flows and can be used to enrich
the context around a specific sample.

Parse these fields from the sample and add them to the FlowMessage.

Signed-off-by: Adrian Moreno <amorenoz@redhat.com>

### CODE REVIEW

File: pb/flow.proto [hunk 1]
```diff
@@ -100,6 +100,9 @@ message FlowMessage {
   uint32 MPLSLastTTL = 61; // Last TTL
   uint32 MPLSLastLabel = 62; // Last Label
 
+  uint32 ObservationDomainID= 70;
```
Comment (lspgn) [inline pb/flow.proto:103]: Missing space

**Comment by lspgn** [comment]
Hello,

> enrich the context around a specific sample.
Could you give me examples on how it's used?

Please bear in mind: I prefer avoiding changing the proto format when:
* it's only on one protocol (nothing on sFlow side)
* I have not seen requests for it, to keep a one size fits all and avoid overloading the samples

**Comment by amorenoz** [comment]
Hi!

Thanks for the quick review.

> Hello,
> 
> > enrich the context around a specific sample.
> > Could you give me examples on how it's used?
> 

The ObservationDomainID and ObservationPointID are used to identify where the sample was taken. Typically each "metering device" has an ObservationDomainID and many ObservationPointIDs. Taking the example from the IPFIX spec: 

```
An Observation Domain is the largest set of Observation Points for
      which Flow information can be aggregated by a Metering Process.
      For example, a router line card may be an Observation Domain if it
      is composed of several interfaces, each of which is an Observation
      Point.  In the IPFIX Message it generates, the Observation Domain
      includes its Observation Domain ID, which is unique per Exporting
      Process.  That way, the Collecting Process can identify the
      specific Observation Domain from the Exporter that sends the IPFIX
      Messages.  Every Observation Point is associated with an
      Observation Domain. 
```

So these identifiers are needed to know where the sample comes from. The use of `SamplerAddress` is not enough because:

- Depending on the network topology, packets from different Samplers might come from the same IP address (e.g: NATing) 
- There can be more than one Exporting Process per "host".
- There can be more than one ObservationPoint per Exporting Process.

As an example, OVS uses the ObservationDomainID and ObservationPointID to provide scope to the samples it generates:

https://www.mankier.com/7/ovs-actions#Other_Actions-The_sample_action
https://www.mankier.com/5/ovs-vswitchd.conf.db#Ipfix_Table

Furthermore, in OVN we're working on some features to [add per-flow sampling](https://patchwork.ozlabs.org/project/ovn/patch/20220613161054.2896553-2-amorenoz@redhat.com/). OVN implements many Logical Datapaths (e.g: Logical Switches, Logical Routers, etc) on top of a single OVS bridge. Each of these datapaths has a "pipeline" with many "stages" that implement different functionalities (e.g: L2 security, MAC learning, Routing, ACL, etc). In this case, we're considering adding an Observation Domain ID per datapath and an observationPointID per stage in the pipeline or "logical flow" so that the IPFIX collector can then understand much more about the sample.

> Please bear in mind: I prefer avoiding changing the proto format when:
> 
>     * it's only on one protocol (nothing on sFlow side)

Unfortunately, these are IPFIX-only attributes.

>     * I have not seen requests for it, to keep a one size fits all and avoid overloading the samples

Sorry, maybe I should have written an issue explaining the reason behind this request.
Maybe we can make it optional?

**Comment by lspgn** [comment]
Thank you for taking the time to write this detailed response! Apologies for taking so long
I understand better the need for this attribute. I added custom fields via config before but only when it's a template field (wouldn't work here) but it makes sense to add it as part of the proto. Will test it and try to merge by this weekend.
Could you just fix the space issue?

> Sorry, maybe I should have written an issue explaining the reason behind this request.

No worries!

**Comment by amorenoz** [comment]
Hi @lspgn , sorry I've been swamped recently with other tasks. I'll update the space.

File: pb/flow.proto (hunk 1 above)
Comment (amorenoz) [reply pb/flow.proto:103]: Done. Thanks!

**Pull request merged**